### DISCUSSION TITLE
Pass on a message from a channel's response transformer to another channel

### DISCUSSION THREAD

**Discussion by Coooz** [nextgenhealthcare/connect#6226] [Q&A] [answered]
In a channel's Source Transformer, I put an incoming HL7 2.4 message on the channel map, to be picked up by a response transformer later on:

`channelMap.put('original_msg', msg);`

Now I noticed that somehow the entire message got embedded in a `<HL7Message>` tag. If I remove those when trying to route the message to another channel, I get a Response Transformer error:

```
154: globalMap.put('fhir_pat_id', fhir_pat_id);
155: //logger.info(channelMap.get('original_msg'));
156: var org_msg = channelMap.get('original_msg');
157: org_msg = org_msg.replace('<HL7Message>', '');
158: org_msg = org_msg.replace('</HL7Message>', '');
159: router.routeMessageByChannelId('25e97ebe-f9be-4855-ae2d-bbc71fd95845', org_msg); //FHIR step 2 ADT Retrieve Observations
160: if ('xml' === typeof msg) {
161:     if (msg.hasSimpleContent()) {
LINE NUMBER:	157
DETAILS:	TypeError: INVALID_CHARACTER_ERR: An invalid or illegal XML character is specified. 
```

However, if I leave them, I get:

```
ER7Serializer error
ERROR MESSAGE: Error converting XML to ER7
org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
	at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source)
	at org.apache.xerces.util.ErrorHandlerWrapper.fatalError(Unknown Source)
```

which somehow seems worse. 

Ultimately, what I need to do is route a HL7 2.4 message (stored on a channel map) from a channel's response transformer to another channel. What is the correct way to do this? I use Mirth Connect 4.4.2.

Thanks!

**Comment by pacmano1** [comment]
Firstly, you should probably stop doing `globalMap.put('fhir_pat_id', fhir_pat_id);`.  It is not thread safe.  I suspect that id is message level concept, so use $c, not $g.  If you need to pass that downstream to other channels, that is what message metadata is for, or encapsulate the forwarded message in something like JSON and pull it apart in the downstream channel.

If you need to store the original HL7 message, get its raw version, not msg.   Or serialize back ER7 (i.e. the pipe "text" version) the xml version you are storing.

But again, what are you _really_ doing here in this channel?

**Reply by Coooz** [reply]
Hi @pacmano1,

What I am _really_ doing is the following: I need to add FHIR data to an incoming HL7 message - I need to do this twice, from different FHIR sources - and then pass the updated message on to an application. Since the FHIR data comes available in the channel's response, I need to have the original message there as well, in order to be able to update this message and forward it to the next channel. That's only in a nutshell, if you like to have more details then I can of course provide them.

I have indeed tried to do away with `globalMap.put('fhir_pat_id', fhir_pat_id);`. That is one of the items I need to enrich the message with, and I have decided to do that right away as soon as `fhir_pat_id` is retrieved (instead of postponing all of the enriching until after all of the FHIR data is available):

```
var org_msg = channelMap.get('original_msg');

var newPID3 = <PID.3/>;
newPID3['PID.3.1'] = fhir_pat_id;
newPID3['PID.3.2'] = '';
newPID3['PID.3.3'] = '';
newPID3['PID.3.4'] = '';
newPID3['PID.3.5'] = 'MR'; //medical record number
org_msg['PID']['PID.3'] += newPID3; //*

channelMap.put('original_msg', org_msg);
```

This code however, does not work when I store the raw version of the original HL7 message, which I do with `channelMap.put('original_msg', connectorMessage.getRawData());` in the Source Transformer. I get 'TypeError: Cannot read property "PID.3" from undefined' in the line marked with * although evaluating `org_msg` when debugging nicely reveals the original message.

![image](https://github.com/nextgenhealthcare/connect/assets/88899312/81e4eae6-931e-4d11-bf17-66d56ff2baca)

So, thinking the above code snippet obviously cannot handle 'the pipe "text" version' I returned to the XML format:
`channelMap.put('original_msg', SerializerFactory.getSerializer('HL7V2').toXML(connectorMessage.getRawData()));`
This got me XML with the `<?xml version="1.0" encoding="UTF-8"?>` header and the already mentioned `<HL7Message>` tag, and resulted in the same 'TypeError: Cannot read property "PID.3" from undefined' in the same line. Removing the header and the tags made no difference.

I am probably overlooking something obvious, both here and in my original question, since I cannot imagine that this should be so difficult. Any further suggestions?

**Edit** The original message has a PID segment, by the way (and a PID.3).

**Reply by pacmano1** [reply]
So to be clear - You're simply receiving an HL7 message and need to supplement that message with a bit of FHIR data from a FHIR http call or two?   Possibly forwarding to another channel?

**Reply by Coooz** [reply]
Yes indeed.

**Reply by Coooz** [reply]
Not "possibly" - I need to forward it to another channel.

**Reply by pacmano1** [reply]
https://github.com/pacmano1/Mirth-Snippets/blob/main/HL7%20add%20FHIR.xml

Import to 4.5 engine and looks at the response transformers for each step.  Stop using $g like that, seriously.  Use $c.

**Reply by pacmano1** [reply, marked as answer]
That example channel shows some response transformer concepts you can utilize.   You could of course just make the FHIR calls and in the final response transformer do everything, because the results of destinations are available downstream as you can see from the examples.   That would mean make the first FHIR call (outbound template would be set to datatype JSON and left empty) and the second destination response transformer:

```JS
tmp  =  new XML(SerializerFactory.getSerializer('HL7V2').toXML(connectorMessage.getRawData())) // this is the original HL7 Message
tmp['PID']['PID.3']['PID.3.1'] = JSON.parse($('d1').getMessage()).id // This is the response from the first destination, getting the "id" from the JSON
tmp['PID']['PID.5']['PID.5.1'] = msg['name'][0]['family'] // This is from the current response 
```

**Reply by Coooz** [reply]
Hi @pacmano1,

Yes, this helped a lot. Thanks for the info!
